fms/modules/tp.py

```diff
@@ -0,0 +1,84 @@
+import itertools
+from abc import ABCMeta, abstractmethod
+from typing import List, Type
+
+import torch
+import torch.nn as nn
+from torch.distributed.distributed_c10d import ProcessGroup
+
+from fms.distributed.tensorparallel import (
+    apply_colwise_tp,
+    apply_embedding_tp,
+    apply_rowwise_tp,
+)
+
+
+class TPModule(nn.Module, metaclass=ABCMeta):
+    """
+    This is an abstract class that any nn.Module can implement to enable
+    Tensor Parallel. On top of inheriting from this class, the TP module
+    will have to implement list_colwise_weights, list_rowwise_weights,
+    list_embedding_weights, and import_module for their relevant weights.
+    Finally, the module must call setup_tp at the end of their __init__
+    function. See examples in attention.py, feedforward.py and embedding.py
+
+    """
+
+    rank: int
+    world_size: int
+
+    def setup_tp(self, rank: int, world_size: int) -> None:
+        self.rank = rank
+        self.world_size = world_size
+
+    def list_colwise_weights(self) -> List[str]:
```

Comment:
I might prefer something like `colwise_param_names` ... to me `list` seems unnecessary and I tend to prefer "params" over "weights" but up to you, not a strong preference